### PR TITLE
T4812: Add op-mode Show vpn ipsec connections

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

Add op-mode CLI `show vpn ipsec connections`
Add the ability to show all configured connections/tunnels and their states.
Ability to get `--raw` data

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4812

## Component(s) name

vpn, ipsec
## How to test
VyOS configuration:
```
set vpn ipsec esp-group ESP-group lifetime '1800'
set vpn ipsec esp-group ESP-group mode 'tunnel'
set vpn ipsec esp-group ESP-group pfs 'enable'
set vpn ipsec esp-group ESP-group proposal 1 encryption 'aes256'
set vpn ipsec esp-group ESP-group proposal 1 hash 'sha256'
set vpn ipsec ike-group IKE-group key-exchange 'ikev1'
set vpn ipsec ike-group IKE-group lifetime '3600'
set vpn ipsec ike-group IKE-group proposal 1 encryption 'aes256'
set vpn ipsec ike-group IKE-group proposal 1 hash 'sha256'
set vpn ipsec interface 'eth1'
set vpn ipsec site-to-site peer OFFICE-B authentication local-id '192.0.2.1'
set vpn ipsec site-to-site peer OFFICE-B authentication mode 'pre-shared-secret'
set vpn ipsec site-to-site peer OFFICE-B authentication pre-shared-secret 'SSSeeccRetT'
set vpn ipsec site-to-site peer OFFICE-B authentication remote-id '192.0.2.2'
set vpn ipsec site-to-site peer OFFICE-B connection-type 'initiate'
set vpn ipsec site-to-site peer OFFICE-B ike-group 'IKE-group'
set vpn ipsec site-to-site peer OFFICE-B local-address '192.0.2.1'
set vpn ipsec site-to-site peer OFFICE-B remote-address '192.0.2.2'
set vpn ipsec site-to-site peer OFFICE-B tunnel 0 esp-group 'ESP-group'
set vpn ipsec site-to-site peer OFFICE-B tunnel 0 local prefix '192.168.0.0/24'
set vpn ipsec site-to-site peer OFFICE-B tunnel 0 remote prefix '10.0.0.0/21'
set vpn ipsec site-to-site peer OFFICE-B tunnel 1 esp-group 'ESP-group'
set vpn ipsec site-to-site peer OFFICE-B tunnel 1 local prefix '192.168.1.0/24'
set vpn ipsec site-to-site peer OFFICE-B tunnel 1 remote prefix '10.0.0.0/21'
set vpn ipsec site-to-site peer OFFICE-B tunnel 2 esp-group 'ESP-group'
set vpn ipsec site-to-site peer OFFICE-B tunnel 2 local prefix '192.168.2.0/24'
set vpn ipsec site-to-site peer OFFICE-B tunnel 2 remote prefix '10.0.0.0/21'
set vpn ipsec site-to-site peer OFFICE-C authentication local-id 'Left-branch'
set vpn ipsec site-to-site peer OFFICE-C authentication mode 'pre-shared-secret'
set vpn ipsec site-to-site peer OFFICE-C authentication pre-shared-secret 'SSSeeccRetT'
set vpn ipsec site-to-site peer OFFICE-C authentication remote-id 'VyOS'
set vpn ipsec site-to-site peer OFFICE-C connection-type 'initiate'
set vpn ipsec site-to-site peer OFFICE-C ike-group 'IKE-group'
set vpn ipsec site-to-site peer OFFICE-C local-address '192.0.2.1'
set vpn ipsec site-to-site peer OFFICE-C remote-address '203.0.113.58'
set vpn ipsec site-to-site peer OFFICE-C tunnel 0 esp-group 'ESP-group'
set vpn ipsec site-to-site peer OFFICE-C tunnel 0 local prefix '192.168.5.0/24'
set vpn ipsec site-to-site peer OFFICE-C tunnel 0 local prefix '192.168.11.0/24'
set vpn ipsec site-to-site peer OFFICE-C tunnel 0 remote prefix '10.0.0.0/21'
set vpn ipsec site-to-site peer OFFICE-C tunnel 0 remote prefix '10.0.10.0/24'

```
show IPsec sa (current exist command):
```
vyos@r14:~$ show vpn ipsec sa
Connection         State    Uptime    Bytes In/Out    Packets In/Out    Remote address    Remote ID    Proposal
-----------------  -------  --------  --------------  ----------------  ----------------  -----------  ---------------------------------------
OFFICE-B-tunnel-0  down     26m28s    0B/0B           0/0               192.0.2.2         192.0.2.2    AES_CBC_256/HMAC_SHA2_256_128/MODP_1024
OFFICE-B-tunnel-0  up       8m1s      0B/0B           0/0               192.0.2.2         192.0.2.2    AES_CBC_256/HMAC_SHA2_256_128/MODP_1024
vyos@r14:~$ 
```
show connections (new command):
```
vyos@r14:~$ show vpn ipsec connections 
Connection         State    Type    Remote address    Local TS         Remote TS     Local id     Remote id    Proposal
-----------------  -------  ------  ----------------  ---------------  ------------  -----------  -----------  ---------------------------------------
OFFICE-B           up       IKEv1   192.0.2.2         -                -             192.0.2.1    192.0.2.2    AES_CBC/256/HMAC_SHA2_256_128/MODP_1024
OFFICE-B-tunnel-0  up       IPsec   192.0.2.2         192.168.0.0/24   10.0.0.0/21   192.0.2.1    192.0.2.2    AES_CBC/256/HMAC_SHA2_256_128/MODP_1024
OFFICE-B-tunnel-1  down     IPsec   192.0.2.2         192.168.1.0/24   10.0.0.0/21   192.0.2.1    192.0.2.2    -
OFFICE-B-tunnel-2  down     IPsec   192.0.2.2         192.168.2.0/24   10.0.0.0/21   192.0.2.1    192.0.2.2    -
OFFICE-C           down     IKEv1   203.0.113.58      -                -             Left-branch  VyOS         -
OFFICE-C-tunnel-0  down     IPsec   203.0.113.58      192.168.5.0/24   10.0.0.0/21   Left-branch  VyOS         -
                                                      192.168.11.0/24  10.0.10.0/24
vyos@r14:~$
```
show connections in `raw` format:
```
vyos@r14:~$ sudo /usr/libexec/vyos/op_mode/ipsec.py show_connections --raw
[
    {
        "ike_connection_name": "OFFICE-B",
        "ike_connection_state": "up",
        "ike_remote_address": [
            "192.0.2.2"
        ],
        "ike_proposal": {
            "cipher": "AES",
            "mode": "CBC",
            "key_size": "256",
            "hash": "HMAC_SHA2_256_128",
            "dh": "MODP_1024"
        },
        "local_id": "192.0.2.1",
        "remote_id": "192.0.2.2",
        "version": "IKEv1",
        "children": [
            {
                "name": "OFFICE-B-tunnel-0",
                "state": "up",
                "local_ts": [
                    "192.168.0.0/24"
                ],
                "remote_ts": [
                    "10.0.0.0/21"
                ],
                "dpd_action": "clear",
                "close_action": "clear",
                "sa": {
                    "name": "OFFICE-B-tunnel-0",
                    "uniqueid": "10",
                    "reqid": "1",
                    "state": "INSTALLED",
                    "mode": "TUNNEL",
                    "protocol": "ESP",
                    "spi_in": "c9701730",
                    "spi_out": "cbc1f06c",
                    "encr_alg": "AES_CBC",
                    "encr_keysize": "256",
                    "integ_alg": "HMAC_SHA2_256_128",
                    "dh_group": "MODP_1024",
                    "bytes_in": "0",
                    "packets_in": "0",
                    "bytes_out": "0",
                    "packets_out": "0",
                    "rekey_time": "2955",
                    "life_time": "1155",
                    "install_time": "645",
                    "local_ts": [
                        "192.168.0.0/24"
                    ],
                    "remote_ts": [
                        "10.0.0.0/21"
                    ]
                },
                "esp_proposal": {
                    "cipher": "AES",
                    "mode": "CBC",
                    "key_size": "256",
                    "hash": "HMAC_SHA2_256_128",
                    "dh": "MODP_1024"
                }
            },
            {
                "name": "OFFICE-B-tunnel-1",
                "state": "down",
                "local_ts": [
                    "192.168.1.0/24"
                ],
                "remote_ts": [
                    "10.0.0.0/21"
                ],
                "dpd_action": "clear",
                "close_action": "clear",
                "sa": {},
                "esp_proposal": {}
            },
            {
                "name": "OFFICE-B-tunnel-2",
                "state": "down",
                "local_ts": [
                    "192.168.2.0/24"
                ],
                "remote_ts": [
                    "10.0.0.0/21"
                ],
                "dpd_action": "clear",
                "close_action": "clear",
                "sa": {},
                "esp_proposal": {}
            }
        ]
    },
    {
        "ike_connection_name": "OFFICE-C",
        "ike_connection_state": "down",
        "ike_remote_address": [
            "203.0.113.58"
        ],
        "ike_proposal": {},
        "local_id": "Left-branch",
        "remote_id": "VyOS",
        "version": "IKEv1",
        "children": [
            {
                "name": "OFFICE-C-tunnel-0",
                "state": "down",
                "local_ts": [
                    "192.168.5.0/24",
                    "192.168.11.0/24"
                ],
                "remote_ts": [
                    "10.0.0.0/21",
                    "10.0.10.0/24"
                ],
                "dpd_action": "clear",
                "close_action": "clear",
                "sa": {},
                "esp_proposal": {}
            }
        ]
    }
]
vyos@r14:~$ 
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: src/op_mode/ipsec.py [hunk 1]
```diff
@@ -135,6 +135,225 @@ def _get_formatted_output_sas(sas):
...
+    """
+    for sa in data:
+        # check if parent SA exist
+        if connection_name not in sa.keys():
+            return '-'
+        encr_alg = sa.get(connection_name, '').get('encr-alg')
+        encr_keysize = sa.get(connection_name, '').get('encr-keysize')
+        integ_alg = sa.get(connection_name, '').get('integ-alg')
+        dh_group = sa.get(connection_name, '').get('dh-group')
+        proposal = f'{encr_alg}/{encr_keysize}/{integ_alg}/{dh_group}'
```
Comment (dmbaturin) [inline src/op_mode/ipsec.py]: I believe this formatting should be done _only_ in human-readable formatted output. The raw output show include the cipher, the key size, and the hash as separate fields. Well, the DH group when applicable, as well, as a separate field.

If we do it like this and send monolithic strings to clients, clients will _parse_ it to display proposal details in GUIs etc. and that particular format will become de facto a public API contact. We most definitely don't want that. ;)

Ideally, we should even separate the cipher and its mode, like this:

```
{
  "cipher": "aes",
  "mode": "cbc",
  "key_size": 256,
  "hash": "HMAC_SHA2_256_128",
  "dh": "MODP_1024"
}
```

We may even want to convert `HMAC_SHA2_256_128` to more conventional names like `SHA-256` (or, maybe, `SHA2-256`), unless there are compelling reasons not to. I have to admit I can't remember what's 128 in that example and I'll need to find out. But I'm sure in an overview in a GUI or similar, people will want to see it as `SHA-256`.

File: src/op_mode/ipsec.py [hunk 2]
```diff
@@ -135,6 +135,225 @@ def _get_formatted_output_sas(sas):
...
+        return child_sa_info[-1] if child_sa_info else {}
+
+
+def _get_child_sa_proposal(child_sa_data: dict) -> str:
+    if 'encr-alg' in child_sa_data:
+        encr_alg = child_sa_data.get('encr-alg')
+        encr_keysize = child_sa_data.get('encr-keysize')
+        integ_alg = child_sa_data.get('integ-alg')
+        dh_group = child_sa_data.get('dh-group')
+        proposal = f'{encr_alg}/{encr_keysize}/{integ_alg}/{dh_group}'
```
Comment (dmbaturin) [inline src/op_mode/ipsec.py]: This should remain machine-readable in the raw output just like the parent SA proposal data.

**Review by dmbaturin — CHANGES_REQUESTED** [review]

File: src/op_mode/ipsec.py (hunk 2 above)
Comment (sever-sever) [reply src/op_mode/ipsec.py]: done!

File: src/op_mode/ipsec.py (hunk 1 above)
Comment (sever-sever) [reply src/op_mode/ipsec.py]: So if we want to convert we need all possible values that we want
or just show it as it is.

**Review by jestabro — APPROVED** [review]

**Review by dmbaturin — APPROVED** [review]

**Pull request merged**